src/data-fetchers/bam/bam-worker.ts

```diff
@@ -228,78 +229,65 @@ const bamRecordToJson = (bamRecord: BamRecord, chrName: string, chrOffset: numbe
 
 type JsonBamRecord = ReturnType<typeof bamRecordToJson>;
 
-// promises indexed by urls
-const bamFiles: Record<string, BamFile> = {};
-const bamHeaders: Record<string, ReturnType<BamFile['getHeader']>> = {};
+class BamFile extends _BamFile {
+    headerPromise: ReturnType<BamFile['getHeader']>;
+    constructor(...args: ConstructorParameters<typeof _BamFile>) {
+        super(...args);
+        this.headerPromise = this.getHeader();
+    }
+    static fromUrl(url: string, indexUrl: string) {
+        return new BamFile({
+            bamFilehandle: new RemoteFile(url),
+            baiFilehandle: new RemoteFile(indexUrl)
+            // fetchSizeLimit: 500000000,
+            // chunkSizeLimit: 100000000,
+            // yieldThreadTime: 1000,
+        });
+    }
+}
 
-const MAX_TILES = 20;
+interface BamFileOptions {
+    loadMates: boolean;
+    maxInsertSize: number;
+    extractJunction: boolean;
+    junctionMinCoverage: number;
+}
 
+// indexed by dataset uuid
+const dataSources: Map<string, DataSource<BamFile, BamFileOptions>> = new Map();
+// indexed by bam url
+const bamFileCache: Map<string, BamFile> = new Map();
+const MAX_TILES = 20;
 const tileValues = new QuickLRU<string, JsonBamRecord[] | { error: string }>({ maxSize: MAX_TILES });
 
-export type DataConfig = {
-    bamUrl: string;
-    baiUrl: string;
-    chromSizes: [string, number][];
-    loadMates?: boolean;
-    maxInsertSize?: number;
-    extractJunction?: boolean;
-    junctionMinCoverage?: number;
-};
-
-type HiGlassDataConfig = Omit<DataConfig, 'baiUrl' | 'chromSizes'> & {
-    chromInfo: ExtendedChromInfo;
-};
-
-// indexed by uuid
-const dataConfs: Record<string, HiGlassDataConfig> = {};
-
 const init = (
     uid: string,
-    {
-        bamUrl,
-        baiUrl,
-        chromSizes,
-        loadMates = false,
-        maxInsertSize = 5000,
-        extractJunction = false,
-        junctionMinCoverage = 1
-    }: DataConfig
+    bam: { url: string; indexUrl?: string },
```

Comment:
I think we can make `indexUrl` required since it is a required property in BamData ([code](https://github.com/gosling-lang/gosling.js/blob/21edf06b4abdd2f6adcca8d323664147d876ca96/src/core/gosling.schema.ts#L1060)). (This is my fault not to update this when I changed the grammar).

```suggestion
    bam: { url: string; indexUrl: string },
```